### DISCUSSION TITLE
Using a custom error type with Effect.task

### DISCUSSION THREAD

**Discussion by abdulajet** [pointfreeco/swift-composable-architecture#733] [Q&A]
Hi folks, 

I am trying to use async/await with the new `Effect.task` method to load some data. Looking at the source Effect.task either supports a `Never` or `Error` type for the second generic parameter on `Effect`. 

Here is some code: 

```
struct AsyncLoginClient {
    var user: (String) -> Effect<Follow, Error>
}

extension AsyncLoginClient {
    static let live = AsyncLoginClient(
        user: { username in
            Effect.task {
                let url = URL(string: Constants.baseURL + "/user?username=\(username)")!
                let (data, _) = try await URLSession.shared.data(for: URLRequest(url: url))
                return try JSONDecoder().decode(Follow.self, from: data)
            }
        })
}
```

I think this matches the example in the comments above the `Effect.task` source. But then because `Error` doesn't conform to `Equatable` my enum of actions can no longer conform to `Equatable`:

```
enum LoginAction {
    case loginResponse(Result<Follow, Error>)
}
```

I guess I can make the `Follow` on my effect optional and do error handling that way but that would be less than ideal. 

I have got this working:

```
struct LoginClient {
    var user: (String) -> Effect<Follow, Failure>
    
    struct Failure: Error, Equatable {}
}

extension LoginClient {
    static let live = LoginClient(
        user: { username in
            return Effect.task {
                let url = URL(string: Constants.baseURL + "/user?username=\(username)")!
                let (data, _) = try await URLSession.shared.data(for: URLRequest(url: url))
                return try JSONDecoder().decode(Follow.self, from: data)
            }
            .mapError { error in
                return LoginClient.Failure()
            }
            .eraseToEffect()
        })
}
```

But the map swallows up my errors. I guess I can switch to an error enum rather than a struct ?

Is this the recommended?

**Comment by mbrandonw** [comment]
> But the map swallows up my errors.

Can you explain a bit more what you mean by this? It seems to me that this code should do what you want. It turns any `Swift.Error` emitted by the task to a `LoginClient.Failure`, but the publisher still fails. It isn't ignored in anyway.

**Reply by abdulajet** [reply]
Ok so I am on the correct track with the `mapError`.

> Can you explain a bit more what you mean by this?

Sorry, in its current state any throw/error will become a `LoginClient.Failure` So I will want to flesh out my struct and return useful info rather than the empty struct. 

I just wanted to check that I was on the the right path, thank you!

**Reply by mbrandonw** [reply]
Ah yeah I see. You can definitely fill out the struct to hold onto some information from the error, such as the `localizedDescription`. You could also turn it into an enum to differentiate between the two types of errors that are possible your code, which is network failure or decoding failure:

```swift
enum Failure: Error {
  case decoding
  case network
}
```

That would only really be necessary if you really do need to tell the difference between those errors in your feature, which is often not the case.

**Reply by abdulajet** [reply]
Amazing thank you!
